src/Stratis.Bitcoin/Base/ChainRepository.cs

```diff
@@ -17,23 +18,110 @@ public interface IChainRepository : IDisposable
+
+        /// <summary>Loads the finalized block height.</summary>
+        /// <returns>Height of a block that can't be reorged away from.</returns>
+        Task LoadFinalizedBlockHeightAsync();
+
+        /// <summary>Saves the finalized block height.</summary>
+        /// <param name="height">Block height.</param>
+        Task SaveFinalizedBlockHeightAsync(int height);
+    }
+
+    public class ChainRepository : IChainRepository, IFinalizedBlockHeight
+    {
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
         /// <summary>Access to DBreeze database.</summary>
         private readonly DBreezeEngine dbreeze;
 
         private BlockLocator locator;
 
-        public ChainRepository(string folder)
+        /// <summary>Database key under which the block hash of the coin view's current tip is stored.</summary>
+        private static readonly byte[] finalizedBlockKey = new byte[0];
+
+        /// <summary>Height of a block that can't be reorged away from.</summary>
+        private int finalizedHeight;
+
+        public ChainRepository(string folder, ILoggerFactory loggerFactory)
         {
             Guard.NotEmpty(folder, nameof(folder));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
 
             this.dbreeze = new DBreezeEngine(folder);
         }
 
-        public ChainRepository(DataFolder dataFolder)
-            : this(dataFolder.ChainPath)
+        public ChainRepository(DataFolder dataFolder, ILoggerFactory loggerFactory)
+            : this(dataFolder.ChainPath, loggerFactory)
+        {
+        }
+
+        /// <inheritdoc />
+        public int GetFinalizedBlockHeight()
+        {
+            return this.finalizedHeight;
+        }
+
+        /// <inheritdoc />
+        public Task LoadFinalizedBlockHeightAsync()
+        {
+            this.logger.LogTrace("()");
+
+            Task task = Task.Run(() =>
+            {
+                this.logger.LogTrace("()");
+
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    transaction.ValuesLazyLoadingIsOn = false;
+                    
+                    Row<byte[], int> row = transaction.Select<byte[], int>("FinalizedBlock", finalizedBlockKey);
+                    if (!row.Exists)
+                    {
+                        this.finalizedHeight = 0;
+                        this.logger.LogTrace("(-):0");
+                        return;
+                    }
+                    
+                    this.finalizedHeight = row.Value;
+                    this.logger.LogTrace("(-):{0}", row.Value);
+                }
+            });
+
+            this.logger.LogTrace("(-)");
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task SaveFinalizedBlockHeightAsync(int height)
```

Comment:
Add remarks saying that this method does not implement the logic to recognise the value provided is correct and it's just simply saves the value to the database. I think this is a mistake and the logic should be implemented here and not by the user of the code. But if you disagree and think that it should not be implemented here, at least at that, saying that it is really not here as one might expect.

Reply:
No longer holds